File: Heap/Heap.java
package Heap;

import java.util.ArrayList;
import java.util.List;

public class Heap {
    private List<Integer> heap;

    public Heap(){
        this.heap = new ArrayList<>();
    
    }

    public List<Integer> getHeap(){
        return new ArrayList<>(heap); // return a copy of the heap, because we don't want to change the original heap, 
                                        // we want to change the copy of the heap, so we can use the heap again, 
                                        // and we can use the original heap again

    }

    public void insert(int value){
        heap.add(value);
        int current = heap.size() - 1; // the index of the last element
        while (current > 0 && heap.get(current) > heap.get(parentIndex(current))) { // if the current value is greater than the parent value
            swap(current, parentIndex(current));
            current = parentIndex(current); 
        }
    }

    public Integer remove() {
        if(heap.size() == 0) return null;
        if(heap.size() == 1) return heap.remove(0);
        int maxValue = heap.get(0);
        heap.set(0, heap.remove(heap.size() - 1)); // set the first element to the last element, and remove the last element
        sinkDown(0);
        return maxValue;
    }

    private void sinkDown(int index){
        int maxIndex = index; // the index of the max value
        while(true){
            int leftChildIndex = leftChildIndex(index);
            int rightChildIndex = rightChildIndex(index);

            if(leftChildIndex < heap.size() && heap.get(leftChildIndex) > heap.get(maxIndex)){ // if the left child value is greater than the max value
                maxIndex = leftChildIndex;
            }
            if(rightChildIndex < heap.size() && heap.get(rightChildIndex) > heap.get(maxIndex)){ // if the right child value is greater than the max value
                maxIndex = rightChildIndex;
            }
            if (maxIndex != index){  // if the max value is not the current value, 
                swap(index, maxIndex);
                index = maxIndex;
            }else{
                return;
            }

        }
       
    }


    private int leftChildIndex(int index){
        return 2 * index + 1;
    }

    private int rightChildIndex(int index){
        return 2 * index + 2;
    }

    private int parentIndex(int index){
        return (index - 1) / 2;
    }

    private void swap(int index1, int index2){ // swap the value of the two indexes
        int temp = heap.get(index1);
        heap.set(index1, heap.get(index2));
        heap.set(index2, temp);
    }

    public static void main(String[] args) {
        Heap heap = new Heap();
        heap.insert(95);
        heap.insert(75);
        heap.insert(80);
        heap.insert(55);
        heap.insert(60);
        heap.insert(50);
        heap.insert(65);

        System.out.println(heap.getHeap());

        heap.remove();

        System.out.println(heap.getHeap());

        heap.remove();

        System.out.println(heap.getHeap());
    }

    /*
    Big O and Time Complexity of Heap:
    very effective for priority queue, because its always balanced and complete
    Hight of the tree is log(n), so insertion and removal is O(log(n))
     */
    
}

File: Seminar3/Task4/heap.py
class Heap:
    def __init__(self):
        self.heap = []

    def get_heap(self):
        return self.heap.copy()

    def insert(self, value):
        self.heap.append(value)
        current = len(self.heap) - 1

        while current > 0 and self.heap[current] < self.heap[self.parent_index(current)]:
            self.swap(current, self.parent_index(current))
            current = self.parent_index(current)

    def remove(self):
        if len(self.heap) == 0:
            return None
        if len(self.heap) == 1:
            return self.heap.pop(0)

        min_value = self.heap[0]
        self.heap[0] = self.heap.pop()

        self.sink_down(0)

        return min_value

    def sink_down(self, index):
        min_index = index

        while True:
            left_child_index = self.left_child_index(index)
            right_child_index = self.right_child_index(index)

            if left_child_index < len(self.heap) and self.heap[left_child_index] < self.heap[min_index]:
                min_index = left_child_index

            if right_child_index < len(self.heap) and self.heap[right_child_index] < self.heap[min_index]:
                min_index = right_child_index

            if min_index != index:
                self.swap(index, min_index)
                index = min_index
            else:
                return

    def left_child_index(self, index):
        return 2 * index + 1

    def right_child_index(self, index):
        return 2 * index + 2

    def parent_index(self, index):
        return (index - 1) // 2

    def swap(self, index1, index2):
        self.heap[index1], self.heap[index2] = self.heap[index2], self.heap[index1]


# if __name__ == "__main__":
#     min_heap = MinHeap()
#     min_heap.insert(95)
#     min_heap.insert(75)
#     min_heap.insert(80)
#     min_heap.insert(55)
#     min_heap.insert(60)
#     min_heap.insert(50)
#     min_heap.insert(65)

#     print(min_heap.get_heap())

#     min_heap.remove()

#     print(min_heap.get_heap())

#     min_heap.remove()

#     print(min_heap.get_heap())
